src/dma.rs

```diff
@@ -1,164 +1,304 @@
+//!   ROM or SRAM. It's usually used for smaller, very high priority transfers.
+//! * DMA 1 and 2 can transfer from ROM memory. These units are intended to be
+//!   used with the FIFO sound buffers, so that playback of a sound can happen
+//!   smoothly regardless of the current position of the CPU within the program.
+//! * DMA 3 can transfer from ROM, and also into ROM if the game cart supports
+//!   that. This DMA unit is the one usually used for loading graphical data
+//!   from ROM into VRAM.
 //!
-//! You are advised to only use the DMA units to transfer data into VRAM,
-//! PALRAM, OAM, and MMIO controls (eg: the FIFO sound buffers).
+//! ## Safety
 //!
-//! In the future the situation may improve.
+//! Using the DMA units is equivalent to playing around with raw pointers. It
+//! must be handled very carefully, or memory corruption can occur.
+
+use core::ffi::c_void;
+
+use bitfrob::{u16_with_bit, u16_with_region};
+use video::Tile4;
+use voladdress::{Safe, Unsafe, VolRegion};
 
-use crate::macros::{pub_const_fn_new_zeroed, u16_bool_field, u16_enum_field};
+use super::*;
 
-/// Sets the change in destination address after each transfer.
-#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+/// Controls the activity of a DMA unit.
+#[derive(Debug, Clone, Copy, Default)]
+#[repr(transparent)]
+pub struct DmaControl(u16);
+impl DmaControl {
+  /// Makes a new, empty value.
+  #[inline]
+  pub const fn new() -> Self {
+    Self(0)
+  }
+  /// Unwrap the raw bits.
+  #[inline]
+  pub const fn into_u16(self) -> u16 {
+    self.0
+  }
+  /// Sets the DMA destination address control, see [`DmaDestAddr`]
+  #[inline]
+  pub const fn with_dest_addr(self, dest: DmaDestAddr) -> Self {
+    Self(u16_with_region(5, 6, self.0, dest as u16))
+  }
+  /// Sets the DMA source address control, see [`DmaSrcAddr`].
+  ///
+  /// When transferring from ROM, this setting is **ignored** and the DMA unit
+  /// will just act as if `Increment` is set.
+  #[inline]
+  pub const fn with_src_addr(self, src: DmaSrcAddr) -> Self {
+    Self(u16_with_region(7, 8, self.0, src as u16))
+  }
+  /// If the DMA unit should repeat again at the next start timing.
+  #[inline]
+  pub const fn with_repeat(self, repeat: bool) -> Self {
+    Self(u16_with_bit(9, self.0, repeat))
+  }
+  /// If the DMA unit should transfer `u32` data (otherwise it's `u16`)
+  #[inline]
+  pub const fn with_u32_transfer(self, u32: bool) -> Self {
+    Self(u16_with_bit(10, self.0, u32))
+  }
+  /// Sets the start timing of the DMA activity, see [`DmaStart`]
+  #[inline]
+  pub const fn with_start_time(self, start: DmaStart) -> Self {
+    Self(u16_with_region(12, 13, self.0, start as u16))
+  }
+  /// If this DMA unit should send an IRQ after it completes the transfer.
+  #[inline]
+  pub const fn with_irq(self, irq: bool) -> Self {
+    Self(u16_with_bit(14, self.0, irq))
+  }
+  /// If the DMA unit is enabled.
```

Comment:
Perhaps 'should be enabled once written to the relevant DMA register'?

Also worth mentioning that there are up to a few cycles of latency before the DMA starts, although in practice I don't think it's possible for the GBA to usefully touch memory in that time.

Reply:
I added some more notes here. The 2 cycle delay is mentioned on the `DmaStart` type